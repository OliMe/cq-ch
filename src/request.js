import getTransport from './event-transport/get-transport';
import { TYPE_QUERY } from './constants';
import { checkArguments } from './helpers/argument-checker';
import { channelCreatorCfg, requestChannelCfg } from './arguments.cfg';

/**
 * Declares a channel for sent queries in the service interface.
 * @param {Array} types Types of queries for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending queries to channel.
 */
export default function request (types, context) {
  checkArguments([types, context], 'request', channelCreatorCfg);
  return async (query, time = 200) => {
    checkArguments([query, time], 'requestChannel', requestChannelCfg);
    if (!types.includes(query.type)) {
      throw new TypeError('Trying to send query with type not in interface.');
    }
    query = { ...query, context };
    return await new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject('Time to answer exhausted.');
      }, time);
      getTransport(TYPE_QUERY).trigger(query.type, {
        ...query, resolve: (value: any) => {
          clearTimeout(timeout);
          return resolve(value);
        },
      });
    });
  };
}
