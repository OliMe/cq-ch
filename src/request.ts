import getTransport from './event-transport/get-transport';
import { TYPE_QUERY } from './constants';
import { Context, Message, OutputQuery, Types } from './types';
import { checkChannelCreator, checkRequestChannel } from './helpers/argument-checkers';

/**
 * Declares a channel for sent queries in the service interface.
 * @param {Array} types Types of queries for sending.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for sending queries to channel.
 */
export default function request(types: Types, context: Context) {
  checkChannelCreator('request', types, context);
  return async function requestChannel<TResponse>(query: Message<TResponse>, time = 200) {
    checkRequestChannel('requestChannel', query, time);
    if (!types.includes(query.type)) {
      throw new TypeError('Trying to send query with type not in interface.');
    }
    return await new Promise((resolve: (value: TResponse) => void, reject) => {
      const timeout = setTimeout(() => {
        reject(
          `Time to answer exhausted. Message type: ${query.type}, request context: ${context}.`,
        );
      }, time);
      const extendedQuery: OutputQuery<TResponse> = {
        ...query,
        context,
        resolve: (value: TResponse) => {
          clearTimeout(timeout);
          resolve(value);
        },
      };
      getTransport(TYPE_QUERY).trigger(extendedQuery.type, extendedQuery);
    });
  };
}
