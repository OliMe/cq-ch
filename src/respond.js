import { channelCreator, takeChannelCreator } from './helpers/channel-creators';
import { TYPE_QUERY } from './constants';
import { checkArguments } from './helpers/argument-checker';
import { channelCreatorCfg } from './arguments.cfg';

/**
 * Declares a channel for processed queries in the service interface.
 * @param {Array} types Types of queries for processing.
 * @param {string} context Application context e.g. Namespace of command.
 * @return {Function} Function for processing queries from channel.
 */
export default function respond (types, context) {
  checkArguments([types, context], 'respond', channelCreatorCfg);
  const channel = channelCreator(types, context);
  return takeChannelCreator(TYPE_QUERY, channel);
}
