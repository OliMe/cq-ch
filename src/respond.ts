import { channelCreator, takeChannelCreator } from './helpers/channel-creators';
import { TYPE_QUERY } from './constants';
import { Context, OutputQuery, Types } from './types';
import { checkChannelCreator } from './helpers/argument-checkers';

/**
 * Declares a channel for processed queries in the service interface.
 * @param types Types of queries for processing.
 * @param context Application context e.g. Namespace of command.
 * @return Function for processing queries from channel.
 */
export default function respond<TResponse>(types: Types, context: Context) {
  checkChannelCreator('respond', types, context);
  const channel = channelCreator<OutputQuery<TResponse>>(types, context);
  return takeChannelCreator<OutputQuery<TResponse>>(TYPE_QUERY, channel)();
}
