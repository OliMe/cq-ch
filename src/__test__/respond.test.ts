import respond from '../respond';
import request from '../request';

describe('respond', () => {
  it('should create respond channel successfully', () => {
    expect(respond(['test_type'], 'test')).toBeInstanceOf(Function);
  });
  it('should create respond channel unsuccessful', () => {
    // TODO Ignore TS errors because for raw JS we need to test argument validation.
    // @ts-ignore
    const createChannelWithoutArguments = () => respond();
    // @ts-ignore
    const createChannelWithFirstIncorrectArgument = () => respond('not array');
    const createChannelWithFirstArgumentEmptyArray = () => respond([], 'test');
    // @ts-ignore
    const createChannelWithoutSecondArgument = () => respond(['test_type']);
    // @ts-ignore
    const createChannelWithSecondIncorrectArgument = () => respond(['test_type'], 200);
    expect(createChannelWithoutArguments).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstIncorrectArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithFirstArgumentEmptyArray).toThrowErrorMatchingSnapshot();
    expect(createChannelWithoutSecondArgument).toThrowErrorMatchingSnapshot();
    expect(createChannelWithSecondIncorrectArgument).toThrowErrorMatchingSnapshot();
  });
});

describe('channel created by respond function', () => {
  it('should receive queries in order of sending and respond with answers', async () => {
    expect.assertions(4);
    const channel = respond(['first_type', 'second_type'], 'first');
    const requestChannel = request(['first_type', 'second_type'], 'second');
    const firstTestQuery = { type: 'first_type' };
    const secondTestQuery = { type: 'second_type' };
    const firstResult = 'first';
    const secondResult = 'second';
    setTimeout(async () => {
      const receivedSecondResult = await requestChannel(secondTestQuery);
      expect(receivedSecondResult).toBe(secondResult);
    }, 25);
    setTimeout(async () => {
      const receivedFirstResult = await requestChannel(firstTestQuery);
      expect(receivedFirstResult).toBe(firstResult);
    }, 20);
    const firstReceivedQuery = await channel();
    expect(firstReceivedQuery).toEqual({
      ...firstTestQuery,
      context: 'second',
      resolve: expect.any(Function),
    });
    firstReceivedQuery && firstReceivedQuery.resolve(firstResult);
    const secondReceivedQuery = await channel();
    expect(secondReceivedQuery).toEqual({
      ...secondTestQuery,
      context: 'second',
      resolve: expect.any(Function),
    });
    secondReceivedQuery && secondReceivedQuery.resolve(secondResult);
  });
});
