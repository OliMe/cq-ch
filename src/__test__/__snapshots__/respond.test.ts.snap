// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`respond should create respond channel unsuccessful 1`] = `"Failed to execute 'respond': first argument must be an Array with at least one element."`;

exports[`respond should create respond channel unsuccessful 2`] = `"Failed to execute 'respond': first argument must be an Array with at least one element."`;

exports[`respond should create respond channel unsuccessful 3`] = `"Failed to execute 'respond': first argument must be an Array with at least one element."`;

exports[`respond should create respond channel unsuccessful 4`] = `"Failed to execute 'respond': second argument must be a string."`;

exports[`respond should create respond channel unsuccessful 5`] = `"Failed to execute 'respond': second argument must be a string."`;
